dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/interpretation/DefaultInterpretationService.java

```diff
@@ -287,7 +288,8 @@ private void notifySubscribers( Interpretation interpretation, InterpretationCom
         if ( interpretableObjectSchema.isSubscribable() )
         {
             SubscribableObject object = (SubscribableObject) interpretableObject;
-            Set<User> subscribers = new HashSet<>( userService.getUsers( object.getSubscribers() ) );
+            Set<User> subscribers = new HashSet<>(
+                userService.getUsers( firstNonNull( object.getSubscribers(), emptySet() ) ) );
```

Comment:
Unfortunately we will see a bit of these mismatches in a transition phase. I think to follow the idea here we should not fix this usage of  `object.getSubscribers()` but change the class(es) that have the method returning a collection to never return a `null` collection. So most likely what these lack is being initialized with an empty collection. Or is is the result of deserialisation?

Reply:
I thought about it, and I agree it would be better to never return null in such cases. It's a bit complicated in this case though...

In this case, sometimes it returns `null` and other times it returns an empty set. This happens because this column is of type JSON and sometimes it is set to NULL and other times to `[]`, in the database (Sierra Leone, likely on some production DB as well). This could happen for a few reasons, and the code should safely deal with it.

_"have the method returning a collection to never return a `null` collection"_
If one sets the collection to be `null` for any reason, it should still return `null` when we do a get on it. Forcing it to return an empty collection if the respective collection is `null` does not sound 100% consistent to me, but could potentially be done.

_"most likely what these lack is being initialized with an empty collection"_
The referred object has the collection initialized as most of the other entities in the project:
`protected Set<String> subscribers = new HashSet<>()`.

Reply:
Thanks for your concern @jbee. If you think we should fix it in a different way feel free to reach out.

Reply:
It depends. If we want/need a semantic distinction between this collection being null or empty we obviously need to handle both. If null/empty have the same consequences we should change the code so even a null in the database ends up with an empty collection in the database and equally an empty collection in the code is stored as either null or empty in the database. Both would need to be handled. 

Because if we fix it by adding code that handles the null returned form the getter we would have to make the same fix for all getters and we are likely to introduce new bugs in the future as the rule is that getters returning collections do not return null. So if we deem this situation to be special and have it digress from this expectation we should also go ahead and add the according annotation `@CheckForNull` to make this clear to the readers/callers.

Reply:
We don't know if in all cases today or in the future, null and empty will mean the same thing. As you said, it depends on each use case.
If you want to generalize it we will lose the ability to make the distinction. Obviously, we can always add it back by creating a new type (a bit overkill though).

But as I see you really prefer forcing the same semantic, let me check if there is a "good" way to always return empty `[]`.

Reply:
After having a conversation with @jbee, we agreed to add a null check in the setter method and create a new instance of the collection if the given parameter is null.